### PR TITLE
flow: use proper direction tcp session values

### PR DESCRIPTION
**Pull request by ilya-bakhtin**
Although most of the flow properties must be taken reversed when we prepare pseudo packets for the flow with FLOW_DIR_REVERSED flag, the tcp sessions must be taken normally, since they are already swapped.
Without this fix, un-acked packets are not analyzed.

Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:

Describe changes:
-
-
-

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

#suricata-verify-pr:
#suricata-verify-repo:
#suricata-verify-branch:
#suricata-update-pr:
#suricata-update-repo:
#suricata-update-branch:
#libhtp-pr:
#libhtp-repo:
#libhtp-branch:

### CODE REVIEW

**Comment by catenacyber** [comment]
Thanks for this contribution.
Could you provide a pcap ? Or even a suricata-verify test ?

un-acked packets are due to missing ack in the capture, right ?
Are there later acks ?

**Comment by ilya-bakhtin** [comment]
Yes, there are missing acks at the very end of the stream. Reassembly pseudo packets are created exactly to force flow reassembly if the normal finalizing sequence is missing. The issue i'm reporting is a result of the following condition:
- the start of the flow is missing and probing parser detected that flow is reversed
- and the end of the flow is missing so reassembly pseudo packets are created and enqueued

The issue is found by the code analysis and then verified by the debug log inspection.
I attached the pcap file that demonstrates inconsistent values of stream last_ack and base_seq.
(stream-tcp-reassemble.c:1014) <Debug> (ReassembleUpdateAppLayer) -- last_ack 2767083861, base_seq 3296893618
It definitely indicates seq/ask values in reassembly packets taken from the wrong direction.

[http-start-from-response-unfinished.zip](https://github.com/OISF/suricata/files/4982757/http-start-from-response-unfinished.zip)

File: src/flow-timeout.c [hunk 1]
```diff
@@ -221,16 +222,20 @@ static inline Packet *FlowForceReassemblyPseudoPacketSetup(Packet *p,
...
-
         /* to client */
     } else {
         p->tcph->th_sport = htons(f->dp);
         p->tcph->th_dport = htons(f->sp);
+    }
 
-        p->tcph->th_seq = htonl(ssn->server.next_seq);
-        p->tcph->th_ack = htonl(ssn->client.last_ack);
+    if (ssn != NULL) {
```
Comment (catenacyber) [inline src/flow-timeout.c:231]: Could you add some comments here ? (basically what you say in the PR, so it stays in the code)

**Comment by catenacyber** [comment]
@ilya-bakhtin  what is the status on this ?

**Comment by ilya-bakhtin** [comment]
Well, i already forgot the details.
We need some suricata-verify. However, as i remember, i can only propose pcap for the test that requires debug.log inspection. Nothing is visible in the eve.json, etc.
Is the test based on debug.log message suitable?

**Comment by catenacyber** [comment]
> Without this fix, un-acked packets are not analyzed.

Can you craft a S-V test where we have a signature that should match on this un-acked packet ?

**Comment by ilya-bakhtin** [comment]
I've, read the latest history of flow-timeout.c and as i can see Victor made the similar change in commit 49bd1f85b99e8bd473ae28c6eb93510b2b3c7668

That fix does the same as i proposed here with tcp session values.
However, i guess it results in the wrong port values of pseudo packets. I'll research this more carefully. But the s-v test becomes much more tricky since the main issue with seq/ack values is already fixed in Victor's commit.

**Comment by catenacyber** [comment]
Closing this PR, feel free to open a new one if there is still something to fix

**Pull request closed without merging**